src/ispc/kernels/lanczos3.ispc

```diff
@@ -90,43 +92,66 @@ struct SampleWeights {
     uniform const WeightCollection* horizontal_weights;
 };
 
-uint8<3> sample_3_channels(const uniform uint8* varying pixel_ptr) {
+float<3> sample_3_channels(const uniform uint8* varying pixel_ptr, uniform PixelFormat pixel_format) {
     // Memory reinterpretation to read all channels at once rather than one-by-one.
     // While testing, this proved more performant than reading one-by-one.
     const uniform uint8<3>* pixel_ptr3 = (const uniform uint8<3>*)(pixel_ptr);
-    varying uint8<3> dst = {0, 0, 0};
-    dst = *pixel_ptr3;
+    varying float<3> dst = {0.0f, 0.0f, 0.0f};
+    dst = *pixel_ptr3 / 255.0f;
+
+    if (is_snorm(pixel_format)) {
+        dst = dst * 2.0f - 1.0f;
+    }
     return dst;
 }
 
-void clean_and_write_3_channels(varying float<3> color, uniform uint8* varying pixel_ptr) {
+float<4> sample_4_channels(const uniform uint8* varying pixel_ptr) {
+    const uniform uint8<4>* pixel_ptr4 = (const uniform uint8<4>*)(pixel_ptr);
+    varying float<4> dst = {0.0f, 0.0f, 0.0f, 0.0f};
+    dst = *pixel_ptr4 / 255.0f;
+
+    return dst;
+}
+
+void clean_and_write_3_channels(varying float<3> color, uniform uint8* varying pixel_ptr, uniform PixelFormat pixel_format) {
     // The final color is a sum of numbers that are multiplied by the weights of their respective pixels.
     // Because of their numbers, floating point precision leads to the final color being potentially outside of the 0-255 range by a slight margin.
     // This would cause an underflow/overflow, which we avoid with the clamps.
-    pixel_ptr[0] = clamp(color[0], 0.0f, 255.0f);
-    pixel_ptr[1] = clamp(color[1], 0.0f, 255.0f);
-    pixel_ptr[2] = clamp(color[2], 0.0f, 255.0f);
-}
+    if (is_snorm(pixel_format)) {
+        // Assumption: Snorm formats are only used for normal maps, so we normalize them to preserve vector length.
+        color = normalize(color);
+        color = color * 0.5f + 0.5f;
```

Comment:
I can't find the code that does the reverse when unpacking,  also we probably need to know the space of the normal map here (typically tangent space, and we only have 2 components). I don't think this is correct what's happening here.